include/kernels/blas_3d.cuh

```diff
@@ -0,0 +1,341 @@
+    static constexpr int nSpin = 1;
+    static constexpr bool spin_project = false;
+    static constexpr bool spinor_direct_load = false; // false means texture load
+    static constexpr bool disable_ghost = true;
+
+    // Create a typename F for the ColorSpinorFields
+    typedef typename colorspinor_mapper<Float, nSpin, nColor, spin_project, spinor_direct_load, disable_ghost>::type F;
+
+    static constexpr int MAX_ORTHO_DIM = 64;
+    complex<real> a[MAX_ORTHO_DIM];
+    const F x;
+    complex<real> b[MAX_ORTHO_DIM];
+    F y;
+
+    caxpby3dArg(const std::vector<Complex> &a, ColorSpinorField &x, const std::vector<Complex> &b, ColorSpinorField &y) :
+      baseArg(dim3(x.VolumeCB(), x.SiteSubset(), 1), x), x(x), y(y)
+    {
+      if (x.X(3) > MAX_ORTHO_DIM) errorQuda("Orthogonal dimension %d exceeds maximum %d", x.X(3), MAX_ORTHO_DIM);
+      for (auto i = 0u; i < a.size(); i++) this->a[i] = a[i];
+      for (auto i = 0u; i < b.size(); i++) this->b[i] = b[i];
+    }
+  };
+
+  template <typename Arg> struct caxpby3d {
+    const Arg &arg;
+    constexpr caxpby3d(const Arg &arg) : arg(arg) { }
+    static constexpr const char *filename() { return KERNEL_FILE; }
+
+    __device__ __host__ inline void operator()(int x_cb, int parity)
+    {
+      using real = typename Arg::real;
+      using Vector = ColorSpinor<real, Arg::nColor, Arg::nSpin>;
+
+      int idx[4];
+      getCoords(idx, x_cb, arg.X, parity);
+
+      Vector x = arg.x(x_cb, parity);
+      Vector y = arg.y(x_cb, parity);
+
+      // Get the coeffs for this dim slice
+      complex<real> a = arg.a[idx[3]];
+      complex<real> b = arg.b[idx[3]];
+
+      // Compute the caxpby
+      Vector out = a * x + b * y;
+
+      // Write to y
+      arg.y(x_cb, parity) = out;
+    }
+  };
+
+  template <typename Float, int nColor_> struct reDotProduct3dArg : public ReduceArg<double> {
+    using real = typename mapper<Float>::type;
+    static constexpr int reduction_dim = 3; // DMH Template this
+    static constexpr int nColor = nColor_;
+    static constexpr int nSpin = 1;
+    static constexpr bool spin_project = false;
+    static constexpr bool spinor_direct_load = false; // false means texture load
+
+    // Create a typename F for the ColorSpinorFields
+    typedef typename colorspinor_mapper<Float, nSpin, nColor, spin_project, spinor_direct_load>::type F;
+
+    F x;
+    F y;
+    int_fastdiv Xh[4]; // checkerboard grid dimensions
+
+    reDotProduct3dArg(const ColorSpinorField &x, const ColorSpinorField &y) :
+      ReduceArg<double>(dim3(x.VolumeCB() / x.X()[reduction_dim], x.SiteSubset(), x.X()[reduction_dim]),
+                        x.X()[reduction_dim]),
+      x(x),
+      y(y)
+    {
+      for (int i = 0; i < 4; i++) Xh[i] = x.SiteSubset() == 2 && i == 0 ? x.X()[i] / 2 : x.X()[i];
+    }
+
+    __device__ __host__ double init() const { return double(); }
+  };
+
+  template <int reduction_dim, class T> __device__ int idx_from_t_xyz(int t, int xyz, T X[4])
```

Comment:
We should put this into `index_helper.cuh`

Reply:
3767371511982bbe8a6ec4cb23a09dda501b6272